analyzers/src/SonarAnalyzer.CSharp/Rules/RedundantCast.cs

```diff
@@ -18,161 +18,167 @@
-                ? returnType.TypeArguments.Single()
-                : null;
-
-        private static bool CanHaveNullValue(ITypeSymbol type) => type.IsReferenceType || type.Name == "Nullable";
-
-        private static Location GetReportLocation(InvocationExpressionSyntax invocation, bool methodCalledAsStatic)
+        var castTypeInfo = context.SemanticModel.GetTypeInfo(type);
+        if (castTypeInfo.Type is not { } castType)
         {
-            if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
-            {
-                return invocation.Expression.GetLocation();
-            }
+            return;
+        }
 
-            return methodCalledAsStatic
-                ? memberAccess.GetLocation()
-                : memberAccess.OperatorToken.CreateLocation(invocation);
+        if (expressionType.Equals(castType) && FlowStateIsNoneOrMatchesCast(expressionTypeInfo, type))
+        {
+            ReportIssue(context, expression, location, castType);
         }
+    }
+
+    private static bool FlowStateIsNoneOrMatchesCast(TypeInfo expressionTypeInfo, ExpressionSyntax type)
+    {
+        var castingToNullable = type is NullableTypeSyntax;
+        return expressionTypeInfo.Nullability().FlowState switch
+        {
+            NullableFlowState.None => true,
+            NullableFlowState.MaybeNull => castingToNullable,
+            NullableFlowState.NotNull => !castingToNullable,
+            _ => throw new NotSupportedException()
+        };
+    }
 
-        private static ITypeSymbol GetElementType(InvocationExpressionSyntax invocation, IMethodSymbol methodSymbol,
-            SemanticModel semanticModel)
+    private static void CheckExtensionMethodInvocation(SonarSyntaxNodeReportingContext context)
+    {
+        var invocation = (InvocationExpressionSyntax)context.Node;
+        if (GetEnumerableExtensionSymbol(invocation, context.SemanticModel) is { } methodSymbol)
         {
-            ExpressionSyntax collection;
-            if (methodSymbol.MethodKind == MethodKind.Ordinary)
+            var returnType = methodSymbol.ReturnType;
+            if (GetGenericTypeArgument(returnType) is { } castType)
             {
-                if (!invocation.ArgumentList.Arguments.Any())
+                if (methodSymbol.Name == "OfType" && CanHaveNullValue(castType))
                 {
-                    return null;
+                    // OfType() filters 'null' values from enumerables
+                    return;
                 }
-                collection = invocation.ArgumentList.Arguments.First().Expression;
-            }
-            else
-            {
-                if (!(invocation.Expression is MemberAccessExpressionSyntax memberAccess))
+
+                var elementType = GetElementType(invocation, methodSymbol, context.SemanticModel);
+                // Generic types {T} and {T?} are equal and there is no way to access NullableAnnotation field right now
```

Comment:
I don't think this is true anymore. You may want to have a look here as well (if it turns out to be too complicated we may also create an issue and come back to it later).

Reply:
I don't see a straight-forward solution. Let's come back to it later.